src/test/scala/pt/tecnico/dsi/neutron/CrudSpec.scala

```diff
@@ -0,0 +1,73 @@
+package pt.tecnico.dsi.neutron
+
+import cats.effect.IO
+import cats.implicits._
+import io.circe.{Decoder, Encoder}
+import pt.tecnico.dsi.neutron.models.Model
+import pt.tecnico.dsi.neutron.services.{BulkCreate, CrudService}
+import pt.tecnico.dsi.openstack.common.models.WithId
+
+abstract class CrudSpec[T <: Model](val name: String, val service: NeutronClient[IO] => CrudService[IO, T])
+  (implicit val encoder: Encoder[T#Create], val decoder: Decoder[WithId[T#Read]])
+  extends Utils {
+
+  val displayName: String = name.capitalize
+  val createStub: IO[T#Create]
+
+  val withSubCreated: IO[(WithId[T#Read], CrudService[IO, T])] =
+    for {
+      neutron <- client
+      stub <- createStub
+      crudService = service(neutron)
+      createdStub <- crudService.create(stub)
+    } yield (createdStub, crudService)
+
+
+  s"$displayName service" should {
+
+    "create and get" in {
+      for {
+        (stub, service) <- withSubCreated
+        actual <- service.get(stub.id)
+      } yield actual shouldBe stub
+    }
+
+    "delete" in {
+      for {
+        (stub, service) <- withSubCreated
+        _ <- service.delete(stub.id)
+        lst <- service.list().compile.toList
+      } yield assert(!lst.exists(_.id == stub.id))
+    }
+
+    "list" in {
+      for {
+        (stub, service) <- withSubCreated
+        lst <- service.list().compile.toList
+      } yield assert(lst.exists(_.id == stub.id))
+    }
+  }
+}
+
+trait BulkCreateSpec[T <: Model] {
+  self: CrudSpec[T] =>
+
+  val bulkService: NeutronClient[IO] => BulkCreate[IO, T]
+  val n = 5
+
+  s"$displayName service" should {
+    "create in bulk" in {
+      for {
+        neutron <- client
+        stubs <- (1 to n).map(_ => createStub).toList.sequence
+
+        bulk = bulkService(neutron)
+        expected <- bulk.create(stubs)
+
+        lst <- expected.map { x =>
+          service(neutron).get(x.id).map(y => y == x)
+        }.sequence
+      } yield assert(lst.forall(identity))
```

Comment:
I can't understand what this `forall` is doing.

Reply:
`lst` will be a List[Assertion]. I wanted to be sure the assertions are evaluated.

Reply:
I find this code is not conveying its meaning very well. Maybe something like this conveys it better:
```scala
    "create in bulk" in {
      for {
        neutron <- client
        stubs <- List.fill(n)(createStub).sequence
        createdStubs <- bulkService(neutron).create(stubs)
        fetchedStubs <- createdStubs.traverse(t => service(neutron).get(t.id))
      } yield fetchedStubs should contain allOf createdStubs
    }
```

Reply:
Beautiful

Reply:
The code didn't compile, I used something else instead.